v4/supervisor.go

```diff
@@ -0,0 +1,956 @@
+
+	if len(s.servicesShuttingDown) == 0 {
+		return nil
+	} else {
+		report := UnstoppedServiceReport{}
+		for serviceID, serviceWithName := range s.servicesShuttingDown {
+			report = append(report, UnstoppedService{
+				SupervisorPath: []*Supervisor{s},
+				Service:        serviceWithName.Service,
+				Name:           serviceWithName.name,
+				ServiceToken:   ServiceToken{uint64(s.id)<<32 | uint64(serviceID)},
+			})
+		}
+		s.m.Lock()
+		s.unstoppedServiceReport = report
+		s.m.Unlock()
+		return report
+	}
+}
+
+// String implements the fmt.Stringer interface.
+func (s *Supervisor) String() string {
+	return s.Name
+}
+
+// sendControl abstracts checking for the supervisor to still be running
+// when we send a message. This prevents blocking when sending to a
+// cancelled supervisor.
+func (s *Supervisor) sendControl(sm supervisorMessage) error {
+	var doneChan <-chan struct{}
+	s.ctxMutex.Lock()
+	if s.ctx == nil {
+		s.ctxMutex.Unlock()
+		return ErrSupervisorNotStarted
+	}
+	doneChan = s.ctx.Done()
+	s.ctxMutex.Unlock()
+
+	select {
+	case s.control <- sm:
+		return nil
+	case <-doneChan:
+		return ErrTimeout
+	}
+}
+
+/*
+Remove will remove the given service from the Supervisor, and attempt to Stop() it.
+The ServiceID token comes from the Add() call. This returns without waiting
+for the service to stop.
+*/
+func (s *Supervisor) Remove(id ServiceToken) error {
+	sID := supervisorID(id.id >> 32)
+	if sID != s.id {
+		return ErrWrongSupervisor
+	}
+	err := s.sendControl(removeService{serviceID(id.id & 0xffffffff), nil})
+	if err == ErrTimeout {
+		// No meaningful error handling if the supervisor is stopped.
+		return nil
+	}
+	return err
+}
+
+/*
+RemoveAndWait will remove the given service from the Supervisor and attempt
+to Stop() it. It will wait up to the given timeout value for the service to
+terminate. A timeout value of 0 means to wait forever.
+
+If a nil error is returned from this function, then the service was
+terminated normally. If either the supervisor terminates or the timeout
+passes, ErrTimeout is returned. (If this isn't even the right supervisor
+ErrWrongSupervisor is returned.)
+*/
+func (s *Supervisor) RemoveAndWait(id ServiceToken, timeout time.Duration) error {
+	sID := supervisorID(id.id >> 32)
+	if sID != s.id {
+		return ErrWrongSupervisor
+	}
+
+	var timeoutC <-chan time.Time
+
+	if timeout > 0 {
+		timer := time.NewTimer(timeout)
+		defer timer.Stop()
+		timeoutC = timer.C
+	}
+
+	notificationC := make(chan struct{})
+
+	sentControlErr := s.sendControl(removeService{serviceID(id.id & 0xffffffff), notificationC})
+
+	if sentControlErr != nil {
+		return sentControlErr
+	}
+
+	select {
+	case <-notificationC:
+		// normal case; the service is terminated.
+		return nil
+
+	// This occurs if the entire supervisor ends without the service
+	// having terminated, and includes the timeout the supervisor
+	// itself waited before closing the liveness channel.
+	case <-s.ctx.Done():
+		return ErrTimeout
```

Comment:
What happens if this is called without the service being started? 
I suspect you'll have a nil pointer exception?

Reply:
sendControl above it shouldn't be able to succeed until the supervisor is started.